File: 1D_LINEAR_CONVECTION/1D_LINEAR_CONVECTION.cpp
#include<iostream>
#include<iomanip>

using namespace std;

int main()
{
    // Declaration
    int Nx = 21, Nt = 51, i, j;
    double c, dx, dt, XLeft, XRight, Tini, Tend, ULeft, URight;
    double x[Nx], u[Nx], un[Nx]; 

    // Initialization
    c = 1.0;
    XLeft = 0.0;
    XRight = 2.0;
    ULeft = 1.0;
    URight = 1.0;
    Tini = 0.0;
    dx = (XRight - XLeft)/(Nx - 1);
    dt = 0.01;
    Tend = dt * Nt;

    // Velocity Initialization
    for(i = 0; i < Nx; i++)
    {
        x[i] = i * dx;
        // printf("%f\n", x[i]);
    }

    for(i = 0; i < Nx; i++)
    {
        if(x[i] >= 0.5 && x[i] <= 1.0)
        {
            u[i] = 2.0;
        }
        else
        {
            u[i] = 1.0;
        }
        // printf("%f\n", u[i]);
    }

    //Forward in Time | Backward in Space Scheme
    for(j = 1; j < Nt+1; j++)
    {	
	u[0] = ULeft;
        u[Nx-1] = URight;
	un[0] = ULeft;
	un[Nx-1] =URight;

        for(i = 1; i < Nx-1; i++)
        {
	    un[i] = u[i];		
            u[i] = un[i] - c * (dt/dx) * (un[i] - un[i-1]);
        }
    }

    for(i = 0; i < Nx; i++)
        {
           printf("%f\n", u[i]);
	   //std::cout << u[i];
        }
}

File: 1D_LINEAR_CONVECTION/1D_LINEAR_CONVECTION.f90
PROGRAM LINEAR_CONVECTION_1D
        IMPLICIT NONE

        !DECLARATIONS
        INTEGER :: i, j
        INTEGER, PARAMETER :: Nx = 201, Nt = 51*2, WP = 4
        REAL :: c, dt, dx, TINI, TEND, XLEFT, XRIGHT, ULEFT, URIGHT
        REAL :: x(Nx), un(Nx), u(Nx)

        !INITIALIZATIONS
        c = 1.0       ! c < 1.0 for artificial viscosity effects ~ 0.5
        dt = 0.01
        TINI = 0.0
        XLEFT = 0.0
        XRIGHT = 2.0
        ULEFT = 1.0
        URIGHT = 1.0
        TEND = dt * Nt
        dx = (XRIGHT - XLEFT)/(Nx - 1)

        !SPACE GRID
        DO i = 1, Nx
                x(i) = XLEFT + (i - 1) * dx
                !PRINT *, x(i)
                !WRITE (*, '(F8.4)') x(i)
        END DO
        
        !VELOCITY INIT
        DO i = 1, Nx
                IF (x(i) >= 0.5_WP .AND. x(i) <= 1.0_WP) THEN
                        u(i) = 2.0_WP
                ELSE 
                        u(i) = 1.0_WP
                END IF        
                !WRITE (*, '(F8.4)') u(i)
        END DO

        !FORWARD IN TIME | BACKWARD IN SPACE SCHEME
        DO j =1 , Nt
                u(1) = ULEFT
                u(Nx) = URIGHT
                un(1) = ULEFT
                un(Nx) = URIGHT
                
                DO i = 2, Nx-1
                un(i) = u(i)
                u(i) = un(i) - c * (dt/dx) * (un(i) - un(i - 1))
                !WRITE (*, '(F8.4)') un(i)
                END DO
        END DO

        DO i = 1, Nx
                WRITE (*, '(F15.6)') u(i)
        END DO

END PROGRAM LINEAR_CONVECTION_1D
